doc/neps/nep-00XX-extensible-dtypes.rst.md

```diff
@@ -0,0 +1,414 @@
+        * Currently source or destination can be NULL (relevant for objects), object handleling could be tricky, and we may think about limiting them? Also related to the ``move_references`` flag.
+        * (This has overlap or identical to enabling some ufunc chaining) 
+
+Users will initially only be able to define ``CastingImpl`` in a very limited manner, the minimal API for allowing all past features is:
+  * Automatically wrapping of ``copyswap``/``copyswapn`` for simplicity and existing dtypes.
+  * From a contiguous casting function and additional strided input, passing in the dtype instances. 
+
+However, this should be expanded, and we may simply allow downstream users to override the ``get_casting_function`` slot in the future.
+
+
+### Python level interface
+
+To expose these slots to python, automatic wrappers shall be created for
+the slots if they are defined as python functions (under specific names).
+Some slots maybe exposed as capsule objects to:
+  * Allow reusing existing slots from other dtypes while avoiding slowdowns
+  * Replacing slots with faster C versions (including by just in time compilers)
+
+The Python interface may be a second step, and may be limited in many cases.
+However, it is a specific design goal to allow the defintion e.g. of Unit dtypes
+in Python. This will also require similar python wrapping for ``CastingImpl`` and
+``UFuncImpl`` (and resolvers) in general.
+
+The exact API should be an additional NEP. A likely design goal will be to allow defining new dtypes using:
+
+```
+@np.dtype
+class Coordinate:
+   x: np.float64
+   y: np.float64
+   z: np.float64
+```
+
+
+### Notes on Casting and DType Discovery
+
+The design presented here means that DType classes are first class objects and finding the correct DType class always happens first both for coercion from python and when finding the correct ``UFuncImpl`` to call.
+
+For non-flexible DTypes, the second step is trivial, since they have a canonical implementation (if there is only a single instance, that one should be typically used for backward compatibility though). For flexible DTypes a second pass is needed, this is either an ``adjust_dtypes`` step within UFuncs, or ``__discover_descr_from_pyobject__`` when coercing within ``np.array``. For the latter, this generally means a second pass is necessary for flexible dtypes (although it may be possible to optimize that for common cases). In this case the ``__common_instance__`` method has to be used as well.
+
+There is currently an open question whether ``adjust_dtypes`` may require the values in some cases. This is currently *not* strictly necessary (with the exception that ``objarr.astype("S")`` will use coercion rather than casting logic, a special case that needs to remain). It could be allowed by giving ``adjust_dtypes`` the input array in certain cases. For the moment it seems preferable to avoid this, if such a disovery step is required, it will require a helper function::
```

Comment:
There's a good chance I may not have understood everything, but I believe the issues around `np.uint64` will need to use the values (i.e. greater than `np.iinfo('uint64').max` or not) to correctly set the dtype.